src/curl_jolokia.c

```diff
@@ -0,0 +1,1045 @@
+	cjo_attribute_values_t *attributepool;
+	int attribute_pool_used;
+
+	const char *json_key;
+	yajl_len_t json_key_len;
+
+	cjo_expect_token expect;
+
+	_Bool verify_peer;
+	_Bool verify_host;
+	char *cacert;
+	struct curl_slist *headers;
+	char *post_body;
+	size_t post_body_len;
+	_Bool expect_escapes;
+	cdtime_t interval;
+	int timeout;
+
+	CURL *curl;
+	char curl_errbuf[CURL_ERROR_SIZE];
+	const unsigned char *start, *end;
+	cjo_membuffer_t replybuffer;
+	cjo_membuffer_t itembuffer;
+
+	size_t poolsize;
+	char *pool;
+	const char *poolmax;
+	char *pool_ptr;
+
+	yajl_handle yajl;
+	llist_t *bean_configs;
+	int max_attribute_count;
+	int max_value_len;
+};
+typedef struct cjo_s cjo_t; /* }}} */
+
+static void cjo_init_buffer(cjo_membuffer_t *buf, size_t initial_size) {
+	buf->buffer = calloc(1, initial_size);
+	buf->used = 0;
+	buf->size = initial_size;
+}
+
+static void cjo_append_buffer(cjo_membuffer_t *buf, const char *append,
+			      size_t appendlen) {
+	if (buf->used + appendlen + 1 > buf->size) {
+		size_t newsize = buf->size * 2;
+		char *newbuf = calloc(1, newsize);
+		memcpy(newbuf, buf->buffer, buf->used);
+		free(buf->buffer);
+		buf->buffer = newbuf;
+		buf->size = newsize;
+	}
+	memcpy(buf->buffer + buf->used, append, appendlen);
+	buf->used += appendlen;
+}
+
+static void cjo_release_buffercontent(cjo_membuffer_t *buf) {
+	sfree(buf->buffer);
+	buf->buffer = NULL;
+	buf->size = 0;
+	buf->used = 0;
+}
+
+static void cjo_remember_value(cjo_membuffer_t *ShouldBeInHere,
+			       cjo_membuffer_t *BufferHereIfNot,
+			       const char *ptr, size_t len,
+			       const char **SetThisOne,
+			       yajl_len_t *SetThisLen) {
+	if (!((ptr > ShouldBeInHere->buffer) &&
+	      (ptr < ShouldBeInHere->buffer + ShouldBeInHere->used))) {
+		char *tptr = BufferHereIfNot->buffer + BufferHereIfNot->used;
+		cjo_append_buffer(BufferHereIfNot, ptr, len);
+		ptr = tptr;
+	}
+
+	*SetThisOne = ptr;
+	*SetThisLen = (yajl_len_t) len;
+}
+
+/*
+ * llist lookup methods
+ */
+static int cjo_cfg_compare_attribute(llentry_t *match, void *vdb) {
+	cjo_t *db = (cjo_t *)vdb;
+	cjo_attribute_t *pcfg = (cjo_attribute_t *)match->value;
+
+	if (pcfg->attribute_match_len != db->curr_attribute->json_name_len)
+		return -1;
+	else
+		return strncmp(pcfg->attribute_match, db->curr_attribute->json_name,
+			       db->curr_attribute->json_name_len);
+}
+static int cjo_cfg_compare_bean(llentry_t *match, void *vdb) {
+	cjo_t *db = (cjo_t *)vdb;
+	cjo_bean_t *pbeancfg = (cjo_bean_t *)match->value;
+
+	if (pbeancfg->mbean_match_len != db->json_key_len)
+		return -1;
+	return strncmp(pbeancfg->mbean_match, db->json_key, db->json_key_len);
+}
+
+/*
+ * Aligning data with Configs & Sending it
+ */
+
+static int cjo_get_type(cjo_attribute_t *attribute) {
+	if (attribute->type != NULL) {
+		return -EINVAL;
+	}
+
+	const data_set_t *ds = plugin_get_ds(attribute->type);
+	if (ds == NULL) {
+		static char type[DATA_MAX_NAME_LEN] = "!!!invalid!!!";
```

Comment:
Where is this string literal coming from? I don't see it a second time; how can `attribute->type` be set to this?

Also, in theory `!!!invalid!!!` is a valid type name. If you want to ensure that the string is not a valid type name, include a slash or a hyphen (or both), for example `/--invalid--/`.